third_party/move/move-vm/types/src/value_serde.rs

```diff
@@ -0,0 +1,343 @@
+    Value::struct_(Struct::pack(vec![Value::vector_u8(bytes)]))
+}
+
+fn into_derived_string_struct(sized_id: SizedID) -> Option<Value> {
+    let width = sized_id.serialized_size() as usize;
+
+    // we need to create DerivedString struct that serializes to exactly match given `width`.
+    // I.e: size_u32_as_uleb128(value.len()) + value.len() + size_u32_as_uleb128(padding.len()) + padding.len() == width
+    // As padding has a fixed allowed max width, it is easiest to expand value to have the padding be minimal.
+    // We cannot always make padding to be 0 byte vector (serialized into 1 byte) - as not all sizes are possible
+    // for string due to variable encoding of string length.
+
+    // So we will over-estimate the serialized length of the value a bit.
+    let value_len_width_upper_bound = size_u32_as_uleb128(width - 2); // we subtract 2 because uleb sizes (for both value and padding fields) are at least 1 byte.
+
+    // If we don't even have enough space to store the length of the value, we cannot proceed
+    if width <= value_len_width_upper_bound + 1 {
+        return None;
+    }
+
+    let id_as_string = u64_to_fixed_size_utf8_bytes(
+        sized_id.into(),
+        // fill the string representation to leave 1 byte for padding and upper bound for it's own length serialization.
+        width - value_len_width_upper_bound - 1,
+    )?;
+
+    bytes_and_width_to_derived_string_struct(id_as_string, width)
+}
+
+fn string_to_bytes(value: Struct) -> Option<Vec<u8>> {
+    value
+        .unpack()
+        .ok()?
+        .collect::<Vec<Value>>()
+        .pop()?
+        .value_as::<Vec<u8>>()
+        .ok()
+}
+
+fn derived_string_struct_to_bytes_and_length(value: Struct) -> Option<(Vec<u8>, u32)> {
+    let mut fields = value.unpack().ok()?.collect::<Vec<Value>>();
+    if fields.len() != 2 {
+        return None;
+    }
+    let padding = fields.pop().unwrap().value_as::<Vec<u8>>().ok()?;
+    let value = fields.pop().unwrap();
+    let string_bytes = string_to_bytes(value.value_as::<Struct>().ok()?)?;
+    let string_len = string_bytes.len();
+    Some((
+        string_bytes,
+        u32::try_from(bcs_size_of_byte_array(string_len) + bcs_size_of_byte_array(padding.len()))
+            .ok()?,
+    ))
+}
+
+fn from_utf8_bytes<T: FromStr>(bytes: Vec<u8>) -> Option<T> {
+    String::from_utf8(bytes).ok()?.parse::<T>().ok()
+}
+
+pub struct NativeValueSimpleSerDe;
+
+impl CustomDeserialize for NativeValueSimpleSerDe {
+    fn custom_deserialize<'d, D: Deserializer<'d>>(
+        &self,
+        deserializer: D,
+        _tag: &IdentifierMappingKind,
+        layout: &MoveTypeLayout,
+    ) -> Result<Value, D::Error> {
+        let encoded = match layout {
+            // FIXME Add checks!
+            MoveTypeLayout::U64 => u64::deserialize(deserializer)?,
+            MoveTypeLayout::U128 => u128::deserialize(deserializer)? as u64,
+            layout if is_derived_string_struct_layout(layout) => {
```

Comment:
What if the layout has some other fields along with DerivedString? For example,
```{
     x: u128,
     y: u64,
     z: DerivedString
}```
This seems to be a generalize deserialize function. Shouldn't it be handling all possible cases of having an aggregator inside the layout?

Reply:
Also, here we are only checking if the layout is a DerivedString. What if the layout is an Aggregator or AggregatorSnapshot? Why aren't those cases handled here?

Reply:
1. What if the layout is an Aggregator or AggregatorSnapshot? Why aren't those cases handled here?

For aggregators/snapshots, we repalce the first field, hence matching on u64 and u128. This is not very uniform with what we do for derived strings, but we can make it so later.

2. Note that for 
```
{
     x: u128,
     y: u64,
     z: DerivedString
}
```
you call this code when you encounter `z` because layout will tell you that it is a `DerivedString`. So it is not a general deserialize, but custom for aggregators/snapshots and derived strings

Reply:
> For aggregators/snapshots, we repalce the first field, hence matching on u64 and u128.

For aggregators/snapshots, we replace `ValueImpl::u64/u128` with `ValueImpl::Native`. So, in the match statement during serialization, we will have to consider `ValueImpl::Native` variant as well right?

Reply:
That's right, we do: see in `value_impl.rs` where there is serialization logic for values. I tagged you in the comment there.